clusterlogging installation via OLM

pruan-rht opened the pull request:
Polarion: [OCP-21311](https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-21311).  @anpingli @QiaolingTang, PTAL.  Thanks!

akostadinov commented on features/step_definitions/logging.rb:
```diff
@@ -0,0 +1,128 @@
+### logging related step definitions
+# for 4.x we default logging installation is via OLM only
+
+Given /^logging service has been installed successfully$/ do
```
can we remove old code related to these two steps
```
there should be (\d+) (logging|metrics) services? installed
all deployer logging pods are running in the#{OPT_QUOTED} project
```

pruan-rht replied:
For 4.x I think the file `features/step_definitions/logging_metrics.rb` can be removed.

akostadinov replied:
yes, please remove dead code

anpingli replied:
@pruan-rht   It is better to split this step to cluster-logging operator ready and provide CRD xxx are provided.
Most of cases are as following
1.  Cluster logging operator is ready.
2. Provide different type of resource file or modify the current resource file
So it is very importance to have a Cluster logging operator ready step.

akostadinov replied:
I don't see old file removed. Maybe missed to commit?

pruan-rht replied:
There are some methods in there that might still be applicable.  Since there is currently no conflict, I would prefer to have the removal of the file be done in a separate PR>

akostadinov replied:
ok, but remember to do it ;)

akostadinov commented on features/step_definitions/logging.rb:
```diff
@@ -0,0 +1,128 @@
+  # Create operator group in `openshift-logging` namespace
+  operator_group_yaml = "https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/logging/clusterlogging/OCP-21311/operator-group.yaml"
+  @result = admin.cli_exec(:create, f: operator_group_yaml, n: logging_ns)
+  raise "Error creating namespace" unless @result[:success]
```
maybe another error message

pruan-rht replied:
what do you mean?  A more explicit error message with the name of the namespace?

akostadinov replied:
this error message is same as line 19 but looks like we are not creating a namespace but a group or something

pruan-rht replied:
oh, yeah, copy and paste error...it should say `operatorgroup` .  Good catch!

akostadinov replied:
I still see `namespace` here.

akostadinov commented on features/step_definitions/logging.rb:
```diff
@@ -0,0 +1,128 @@
+  @result = admin.cli_exec(:create, f: catsrc_logging_yaml)
+  raise "Error creating catalogsourceconfig for cluster_logging" unless @result[:success]
+  @result = admin.cli_exec(:create, f: catsrc_elasticsearch_yaml)
+  raise "Error creating catalogsourceconfig for elasticsearch" unless @result[:success]
```
two error messages exactly the same

pruan-rht replied:
one is for `clusterlogging` the other is for  `elasticsearch`

akostadinov commented on features/step_definitions/logging.rb:
```diff
@@ -0,0 +1,128 @@
+  logging_yaml = "https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/logging/clusterlogging/example.yaml"
+  @result = admin.cli_exec(:create, f: logging_yaml, n: logging_ns)
+  # Check the resource created by OLM
+  step %Q/I wait for clusterlogging to be functional in the project/
```
shall we remove what we created at end of scenario? I see step is destructive tagged

pruan-rht replied:
@akostadinov since with OLM install, the logging installation variations is much less than for ansible install for 3.x, I'm thinking for 4.x testing, we just install it once and leave it running.  @anpingli or @QiaolingTang should make a call on that front since they are the owner of those user stories.

anpingli replied:
@pruan-rht  Most of time, we have to remove the clusterlogging/instance 1). it is better to clean it each time.
We needn't to clear the clusterlogging operator/elasticsearch operator and namespace each time.

1) To clean up clusterlogging, just remove the clusterlogging/instance resource  
#oc delete clusterlogging/instance -n openshift-logging
2) To clean up clusterlogging operator, we need to unsubscribe the operator
3) To clean up elasicsearch operator, we need to unsubscribe the operator
     
     
To clean up all resource

pruan-rht replied:
@anpingli Thanks for the feedback.  If you do recommend removing after each install, perhaps we can add an option like `keep_installation: true` to the step and default to remove all previous installations to make it simpler.

anpingli replied:
Looks good to have a option keep_installation: true

akostadinov replied:
Do you intend to make clean-up optional? Or do we drop the idea?

pruan-rht replied:
@akostadinov @anli agrees with having that as optional.

akostadinov replied:
I missed that it is there, sorry.

akostadinov commented on lib/openshift/cluster_logging.rb:
```diff
@@ -0,0 +1,107 @@
+
+    def visualization(user: nil, cached: true, quiet: false)
+      raw_resource(user: user, cached: cached, quiet: quiet).dig('status', 'visualization')
+    end
```
what values do these return?

akostadinov replied:
what I would like to is to have all low-level methods that return some raw YAML to be private. Users shouldn't parse raw structures. Then we can gradually add methods returning proper objects when needed.

Similar to how we have `QuotaLimits#memory_limit_raw` vs `QuotaLimits#memory_limit` methods. The raw method returns unprocessed structure. The non-raw method returns some more convenient object.

pruan-rht replied:
Those are part of the `status` return values.  Unlike `QuotaLimits`, they are not values but rather `yaml/json` structs.  You can look at it here https://github.com/openshift-qe/output_references/blob/master/logging/4.x/clusterlogging.yaml

akostadinov replied:
I know, that's why I'm asking you to name methods `..._raw` and make `private` all that are never called outside the class. We can create deal with any external users when we have such.

pruan-rht commented:
@akostadinov PR udpated.  PTAL again.  Thanks!

akostadinov commented on lib/openshift/cluster_logging.rb:
```diff
@@ -0,0 +1,107 @@
+
+    def fluentd_notready_pods(user: nil, cached: true, quiet: false)
+      self.fluentd_pods(user: user, cached: cached, quiet: quiet)['notReady']
+    end
```
All `fluentd*_pods` methods can be private as far as I can tell. If they are to be public, then they better return `Pod` objects instead of some unknown structure.

akostadinov commented on lib/openshift/cluster_logging.rb:
```diff
@@ -0,0 +1,107 @@
+
+    def kibana_pods(user: nil, cached: true, quiet: false)
+      self.kibana_status(user: user, cached: cached, quiet: quiet).first['pods']
+    end
```
all these methods also return some structure instead of objects user can more easily interact with. And not used outside the class. So `es_cluster_health`, `es_pods`, `kibana_status` and `kibana_pods` can happily be private.

pruan-rht commented:
@akostadinov, I believe I've addressed all of your concerns now, except removal of the file, which I think we should do in another PR because I think some of the methods there are still applicable.  PTAL again.  Thanks!
